torch/distributed/elastic/utils/store.py

```diff
@@ -58,21 +65,48 @@ def synchronize(
 
     Note: The data on the path is not deleted, as a result there can be stale data if
         you use the same key_prefix twice.
+
+    Time complexity: O(N) per worker, O(N^2) globally.
     """
     store.set_timeout(timedelta(seconds=barrier_timeout))
     store.set(f"{key_prefix}{rank}", data)
     agent_data = get_all(store, rank, key_prefix, world_size)
     return agent_data
 
 
+def _barrier_nonblocking(store, world_size: int, key_prefix: str) -> str:
+    """
+    Does all the non-blocking operations for a barrier and returns the final key
+    that can be waited on.
+    """
+    num_members_key = key_prefix + _NUM_MEMBERS
+    last_member_key = key_prefix + _LAST_MEMBER_CHECKIN
+
+
+    idx = store.add(num_members_key, 1)
+    if idx == world_size:
+        store.set(last_member_key, "<val_ignored>")
+
+    return last_member_key
+
+
 def barrier(
-    store, rank: int, world_size: int, key_prefix: str, barrier_timeout: float = 300
+    store, world_size: int, key_prefix: str, barrier_timeout: float = 300
 ) -> None:
     """
-    A global lock between agents.
+    A global lock between agents. This will pause all workers until at least
+    ``world_size`` workers respond.
+
+    This uses a fast incrementing index to assign waiting ranks and a success
+    flag set by the last worker.
+
+    Time complexity: O(1) per worker, O(N) globally.
 
     Note: Since the data is not removed from the store, the barrier can be used
         once per unique ``key_prefix``.
     """
-    data = f"{rank}".encode()
-    synchronize(store, data, rank, world_size, key_prefix, barrier_timeout)
+
+    store.set_timeout(timedelta(seconds=barrier_timeout))
```

Comment:
This changes default timeout, is that OK?

Reply:
That seems to be how these store functions operate -- I don't love it but keeping to existing patterns. We already set it in the existing synchronize call so I expect it's already at that value.

Reply:
Oh yeah -- since this is eliminating synchronize and synchronize was setting the timeout this behavior is exactly the same

Reply:
There will be cases when TCPStore will be shared with and used by trainer via `master_addr/master_port` args. Should we consider scoping it to just this place?

Reply:
Changed to scope it to just this with a context manager -- just a warning that this is a breaking change to previous behavior

Reply:
ideally the store API has `set(key, val, timeout)` and `get(key, timeout)` but unfortuantely that's not the case today. In the case of `TCPStore` this is a client-side timeout so even if we shared the "TCPStore" server between the agent and trainers this timeout won't get propagated down to the trainers.